cpp/inc/bond/core/maybe.h

```diff
@@ -14,141 +19,556 @@ namespace bond
+    {
+        BOOST_ASSERT(!is_nothing());
+        return *_value;
+    }
 
-    /// @brief Swap this object with that object
-    void swap(maybe& that)
+    /// @brief Compares two maybes for value equality.
+    ///
+    /// @return true if both maybes hold nothing; returns false if one maybe
+    /// holds nothing and the other holds a values; otherwise, calls
+    /// operator== with the two values.
+    ///
+    /// @since 8.0.0
+    friend bool operator==(const maybe_common& lhs, const maybe_common& rhs)
     {
-        using std::swap;
+        return lhs._value == rhs._value;
+    }
+
+    /// @brief Compares two maybes for value inequality.
+    ///
+    /// See operator==(const maybe_common&,const maybe_common&) for details
+    /// about how maybes holding nothing are handled.
+    ///
+    /// @since 8.0.0
+    friend bool operator!=(const maybe_common& lhs, const maybe_common& rhs)
+    {
+        return !(lhs == rhs);
+    }
+
+    /// @brief Compares a maybe and a value for equality.
+    ///
+    /// @return false if the maybe holds nothing; otherwise, calls
+    /// operator== with the maybe's value and the provided value.
+    ///
+    /// @since 8.0.0
+    friend bool operator==(const maybe_common& lhs, const T& rhs)
+    {
+        return lhs._value == rhs;
+    }
+
+    /// @brief Compares a maybe and a value for inequality.
+    ///
+    /// See operator==(const maybe_common&,const T&) for details about how
+    /// maybes holding nothing are handled.
+    ///
+    /// @since 8.0.0
+    friend bool operator!=(const maybe_common& lhs, const T& rhs)
+    {
+        return !(lhs == rhs);
+    }
+
+    /// @brief Compares a value and a maybe for equality.
+    ///
+    /// @return false if the maybe holds nothing; otherwise, calls
+    /// operator== with the provided value and maybe's value.
+    ///
+    /// @since 8.0.0
+    friend bool operator==(const T& lhs, const maybe_common& rhs)
+    {
+        return lhs == rhs._value;
+    }
+
+    /// @brief Compares and a value and a maybe for inequality.
+    ///
+    /// See operator==(const T&,const maybe_common&) for details about how
+    /// maybes holding nothing are handled.
+    ///
+    /// @since 8.0.0
+    friend bool operator!=(const T& lhs, const maybe_common& rhs)
+    {
+        return !(lhs == rhs);
+    }
+
+protected:
+    maybe_common() = default;
+
+    #if defined(_MSC_VER) && _MSC_VER < 1900
+    // Using = default with MSVC 2013 on this function causes the compiler
+    // to make extra copies, which makes this type fail when it's holding a
+    // move-only type.
+    //
+    // Manually implementing this function works just fine, however.
+    maybe_common(const maybe_common& that)
+        : _value(that._value)
+    { }
+    #else
+    maybe_common(const maybe_common&) = default;
+    #endif
+
+    template <typename... Args>
+    explicit maybe_common(const T& value, Args&&... args)
+        : _value()
```

Comment:
The `: _value()` is redundant (will not be a POD for any `T`).

Reply:
✔️